compiler/rustc_codegen_llvm/src/va_arg.rs

```diff
@@ -278,6 +278,177 @@ fn emit_s390x_va_arg<'ll, 'tcx>(
     bx.load(val_type, val_addr, layout.align.abi)
 }
 
+fn emit_x86_64_sysv64_va_arg<'ll, 'tcx>(
+    bx: &mut Builder<'_, 'll, 'tcx>,
+    list: OperandRef<'tcx, &'ll Value>,
+    target_ty: Ty<'tcx>,
+) -> &'ll Value {
+    let dl = bx.cx.data_layout();
+
+    // Implementation of the systemv x86_64 ABI calling convention for va_args, see
+    // https://gitlab.com/x86-psABIs/x86-64-ABI (section 3.5.7). This implementation is heavily
+    // based on the one in clang.
+
+    // We're able to take some shortcuts because the return type of `va_arg` must implement the
+    // `VaArgSafe` trait. Currently, only pointers, f64, i32, u32, i64 and u64 implement this trait.
+
+    // typedef struct __va_list_tag {
+    //     unsigned int gp_offset;
+    //     unsigned int fp_offset;
+    //     void *overflow_arg_area;
+    //     void *reg_save_area;
+    // } va_list[1];
+    let va_list_addr = list.immediate();
+
+    let unsigned_int_offset = 4;
+    let ptr_offset = 8;
+    let gp_offset_ptr = va_list_addr;
+    let fp_offset_ptr = bx.inbounds_ptradd(va_list_addr, bx.cx.const_usize(unsigned_int_offset));
+
+    let layout = bx.cx.layout_of(target_ty);
+
+    // AMD64-ABI 3.5.7p5: Step 1. Determine whether type may be passed
+    // in the registers. If not go to step 7.
+
+    // AMD64-ABI 3.5.7p5: Step 2. Compute num_gp to hold the number of
+    // general purpose registers needed to pass type and num_fp to hold
+    // the number of floating point registers needed.
+
+    let mut num_gp_registers = 0;
+    let mut num_fp_registers = 0;
+
+    match layout.layout.backend_repr() {
+        BackendRepr::Scalar(scalar) => match scalar.primitive() {
+            rustc_abi::Primitive::Int(integer, _is_signed) => {
+                num_gp_registers += integer.size().bytes().div_ceil(8) as i32;
+            }
+            rustc_abi::Primitive::Float(float) => {
+                num_fp_registers += float.size().bytes().div_ceil(16) as i32;
+            }
+            rustc_abi::Primitive::Pointer(_) => {
+                num_gp_registers += 1;
+            }
+        },
+        BackendRepr::ScalarPair(..) => {
+            unreachable!("VaArgSafe is not implemented for scalar pairs")
+        }
+        BackendRepr::SimdVector { .. } => {
+            unreachable!("VaArgSafe is not implemented for SIMD vectors")
+        }
+        BackendRepr::Memory { .. } => { /* do nothing */ }
```

Comment:
can these just be `=> unreachable!("No x86-64 SysV va_arg implementation for {:?}", layout.layout.backend_repr())`